dist/jose.js

```diff
@@ -0,0 +1,1710 @@
+  // the Million Message Attack on Cryptographic Message Syntax". We currently
+  // only support RSA-OAEP, so we don't generate a key if unwrapping fails.
+  var encrypted_cek = Utils.Base64Url.decodeArray(parts[1]);
+  var cek_promise = that.key_promise.then(function (key) {
+    return this.cryptographer.unwrapCek(encrypted_cek, key);
+  }.bind(that));
+
+  // part 3: decrypt the cipher text
+  var plain_text_promise = that.cryptographer.decrypt(
+    cek_promise,
+    Utils.arrayFromString(parts[0]),
+    Utils.Base64Url.decodeArray(parts[2]),
+    Utils.Base64Url.decodeArray(parts[3]),
+    Utils.Base64Url.decodeArray(parts[4]));
+
+  return plain_text_promise.then(Utils.stringFromArray);
+};
+
+/*-
+ * Copyright 2015 Peculiar Ventures
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Handles decryption.
+ *
+ * @param cryptographer  an instance of WebCryptographer (or equivalent). Keep
+ *                       in mind that decryption mutates the cryptographer.
+ *
+ * @author Patrizio Bruno <patrizio@desertconsulting.net>
+ */
+JoseJWS.Signer = function (cryptographer) {
+
+  var that = this;
+  that.cryptographer = cryptographer;
+
+  that.key_promises = {};
+  that.waiting_kid = 0;
+  that.headers = {};
+  that.signer_aads = {};
+  that.signer_headers = {};
+};
+
+/**
+ * Adds a signer to JoseJWS instance. It'll be the on of the signers of the resulting JWS.
+ *
+ * @param rsa_key        private RSA key in json format, Parameters can be base64
+ *                       encoded, strings or number (for 'e'). Or CryptoKey
+ * @param key_id         a string identifying the rsa_key. OPTIONAL
+ * @param aad            Object protected header
+ * @param header         Object unprotected header
+ */
+JoseJWS.Signer.prototype.addSigner = function (rsa_key, key_id, aad, header) {
+  var that = this,
+    key_promise = isCryptoKey(rsa_key) ? new Promise(function (resolve) {
+      resolve(rsa_key);
+    }) : Jose.Utils.importRsaPrivateKey(rsa_key, aad && aad.alg ? aad.alg : that.cryptographer.getContentSignAlgorithm(), "sign"),
+    kid_promise;
+
+  if (key_id) {
+    kid_promise = new Promise(function (resolve) {
+      resolve(key_id);
+    });
+  } else if (isCryptoKey(rsa_key)) {
+    throw new Error("key_id is a mandatory argument when the key is a CryptoKey");
+  } else {
+    kid_promise = Jose.WebCryptographer.keyId(rsa_key);
+  }
+
+  that.waiting_kid++;
```

Comment:
The Promise API is powerful enough that this should not be required. Look at `Promise.all` and `Promise.race`.